pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,261 @@
+// Returns a boolean if the deployment exists, and an error when relevant.
+func (r *reconciler) ensureExternalDNSDeployment(ctx context.Context, namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	desired, err := desiredExternalDNSDeployment(namespace, image, serviceAccount, externalDNS)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to build externalDNS deployment: %w", err)
+	}
+
+	exist, current, err := r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to get externalDNS deployment: %w", err)
+	}
+
+	// create the deployment
+	if !exist {
+		if err := r.createExternalDNSDeployment(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		// getting the deployment from API in case a third party updated it after the creation
+		return r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+	}
+
+	// update the deployment
+	if _, err := r.updateExternalDNSDeployment(ctx, current, desired); err != nil {
+		return true, current, err
+	}
+
+	// getting the deployment from API in case a third party updated it after the update
+	return r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
+}
+
+// currentExternalDNSDeployment gets the current externalDNS deployment resource.
+func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, namespace string, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	depl := &appsv1.Deployment{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: externalDNS.Name}, depl); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, depl, nil
+}
+
+// desiredExternalDNSDeployment returns the desired deployment resource.
+func desiredExternalDNSDeployment(namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (*appsv1.Deployment, error) {
+	replicas := int32(1)
+
+	matchLbl := map[string]string{
+		"app.kubernetes.io/name":     "external-dns",
+		"app.kubernetes.io/instance": externalDNS.Name,
+	}
+
+	nodeSelectorLbl := map[string]string{
+		"kubernetes.io/os":               "linux",
+		"node-role.kubernetes.io/master": "",
+	}
+
+	depl := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      controller.ExternalDNSResourceName(externalDNS),
+			Namespace: namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: matchLbl,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: matchLbl,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: serviceAccount.Name,
+					NodeSelector:       nodeSelectorLbl,
+				},
+			},
+		},
+	}
+
+	metricsPort := 7979
+	for i, zone := range externalDNS.Spec.Zones {
```

Comment:
Do we have a sense of how many Zones might there be?  Is is possible that there might be port conflicts if we start venturing into the hundreds of Zones?

Reply:
I don't think that we have an idea about how many zones can be here, in my mind it's quite a small number (1, 2 digits). I think we may have port conflicts in almost any range. It's the proximity to the popular `8000` range which you think is not good? I didn't see externalDNS listening on `8xxx` ports so far.

Reply:
On OpenShift, there will be 1 or 2 zones, and I doubt any practical use cases outside of OpenShift would call for more than a few zones.

Reply:
We should specify `MaxItems` for `spec.zones` in the CRD though.

Reply:
Added `MaxItems` with the value of `10`. @Miciah, let me know which value makes sense here.

Reply:
As per discussion in a chat, let's keep it set to 10 for the moment and we'll see if it'll need a change.